design/ratelimit-design.md

```diff
@@ -0,0 +1,154 @@
+# Envoy Rate Limiting Support
+
+The rate limit service configuration specifies the global rate limit service Envoy should talk to when it needs to make global rate limit decisions.
+If no rate limit service is configured, a “null” service will be used which will always return OK if called.
+
+Envoy's support of RateLimiting requires a rate limiting service to be exposed and support the gRPC IDL specified in [rls.proto](https://www.envoyproxy.io/docs/envoy/v1.9.0/api-v2/service/ratelimit/v2/rls.proto#envoy-api-file-envoy-service-ratelimit-v2-rls-proto).
+
+## Goals
+
+- Allow RateLimits to be applied to routes
+- Use Lyft's [ratelimiting implementation](https://github.com/lyft/ratelimit)
+
+## Non-goals
+
+- Support additional RateLimiting implementations
+- For the initial implementation, do not tightly integrate the RateLimiting service into Contour
+- Contour will not write the RateLimiter configuration (future versions could to make the integration smoother)
+
+## High-level Design
+
+The RateLimiting implementation will determine if the request sent to Envoy should be serviced or not.
+Contour will require configuration to enable the rate limiting HTTP filter as well as pointing Envoy to the rate limit implementation service.
```

Comment:
We should probably only enable the rate limiting filter anywhere in the ingressroute chain has a rate limit stanza.

Reply:
Yes I agree. I think looking at this now I never allowed for a spot to enable the service. That should probably be an arg to Contour as you'd only want to set it up once. Having that existing would be the trigger for contour to enable the filter, etc.